source/base/data_out_base.cc

```diff
@@ -666,51 +666,38 @@ namespace
   }
 
   /**
-   * Class for writing basic
-   * entities in @ref
-   * SoftwareOpenDX format,
-   * depending on the flags.
+   * Class describing common functionality between different output streams.
+   *
+   * @ingroup output
    */
-  class DXStream
+  template<typename FlagsType>
+  class StreamBase
   {
   public:
-    /**
-     * Constructor, storing
-     * persistent values for
-     * later use.
+    /*
+     * Constructor. Stores a reference to the output stream for immediate use.
      */
-    DXStream (std::ostream &stream,
-              const DataOutBase::DXFlags flags);
+    StreamBase (std::ostream &stream, const FlagsType flags)
+      :
+      selected_component (numbers::invalid_unsigned_int),
+      stream (stream),
+      flags (flags)
+    {}
 
     /**
      * Output operator for points.
      */
     template <int dim>
-    void write_point (const unsigned int index,
-                      const Point<dim> &);
+    void write_point (const unsigned int index, const Point<dim> &);
```

Comment:
> I put the declaration there to emphasize that inheriting classes should
> implement that function so that they work with some of the templated functions
> later in the file. I think it is useful, but I can remove it if that is your
> preference.

I find unimplemented but declared functions difficult because it's so easy to 
call them in later changes and so hard to figure out from the linker errors 
what goes wrong :-) I think the documentation aspect is good. How about you 
just give these functions a body of the form
```
   Assert (false, ExcMessage ("The derived class you are using needs to 
re-implement this function if you want to call it."));
```
or similar?